src/mds/Server.cc

```diff
@@ -6253,6 +6261,57 @@ void Server::handle_client_removexattr(MDRequestRef& mdr)
   journal_and_reply(mdr, cur, 0, le, new C_MDS_inode_update_finish(this, mdr, cur));
 }
 
+void Server::handle_client_getvxattr(MDRequestRef& mdr)
+{
+  const cref_t<MClientRequest> &req = mdr->client_request;
+  string name(req->get_path2());
+
+  // is a ceph virtual xattr?
+  if (!is_ceph_vxattr(name)) {
+    respond_to_request(mdr, -CEPHFS_EINVAL);
+    return;
+  }
+
+  // can't use rdlock_path_pin_ref because we need to xlock snaplock/policylock
+  CInode *cur = try_get_auth_inode(mdr, req->get_filepath().get_ino());
+  if (!cur) {
+    return;
+  }
+  if (!cur->is_dir()) {
+    respond_to_request(mdr, -CEPHFS_ENOTDIR);
+    return;
+  }
+
+  // handle these vxattrs
+  int r = 0;
+  ceph::bufferlist bl;
+  if (name == "ceph.dir.layout"sv) {
+    cur->get_inode()->layout.encode(bl, (uint64_t)-1);
```

Comment:
Since these values should eventually be returned from the syscall, I think a text string makes sense. The client/kclient should not need to interpret (i.e. decode) these.

Reply:
as per the discussion, I'm keeping the code to fetch individual layout fields in binary form rather than the proposed text form
this avoids the string parsing hassles on the client side as well

Reply:
Coming back to this: I'm not sure what the point is of binary encoding anything for `getvxattr`. The client (kernel or `Client`) does not need any of these vxattrs to function. If the client **does** need a vxattr, it should be encoded and sent to the client with caps (like file layouts). So I'm not seeing a use-case to binary encode these values for any kind of convenience for the client. I would suggest JSON encoding the values and let the client pass it unmodified to the application/syscall.